examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs

```diff
@@ -0,0 +1,133 @@
+        {
+            // We provide our own custom namespace transformation, and expression kind transformation.
+            this.Namespaces = new ListIdentifiers.NamespaceTransformation(this);
+            this.ExpressionKinds = new ListIdentifiers.ExpressionKindTransformation(this);
+            // Since we only want to modify the QsCallable nodes, there is no need to rebuild statements and expressions.
+            this.Statements = new StatementTransformation<TransformationState>(this, TransformationOptions.NoRebuild);
+            this.StatementKinds = new StatementKindTransformation<TransformationState>(this, TransformationOptions.NoRebuild);
+            this.Expressions = new ExpressionTransformation<TransformationState>(this, TransformationOptions.NoRebuild);
+            // Since we are only interested in all occurring identifiers, we can omit walking type nodes entirely.
+            this.Types = new TypeTransformation<TransformationState>(this, TransformationOptions.Disabled);
+        }
+
+        /// <summary>
+        /// Adds a comment to each callable listing all identifiers used within the callable.
+        /// </summary>
+        public QsCompilation OnCompilation(QsCompilation compilation)
+        {
+            var namespaces = compilation.Namespaces
+                .Select(this.Namespaces.OnNamespace)
+                .ToImmutableArray();
+            return new QsCompilation(namespaces, compilation.EntryPoints);
+        }
+
+        /// <summary>
+        /// Class used to track the internal state of the transformation, as well as access any information based on it.
+        /// These properties and methods are usually used by multiple subtransformations.
+        /// </summary>
+        public class TransformationState
+        {
+            private QsQualifiedName CurrentCallable;
+            private readonly Dictionary<QsQualifiedName, ImmutableHashSet<string>.Builder> Identifiers =
+                new Dictionary<QsQualifiedName, ImmutableHashSet<string>.Builder>();
+
+            public ImmutableHashSet<string> IdentifiersInCallable(QsQualifiedName name) =>
+                this.Identifiers.TryGetValue(name, out var ids) ? ids.ToImmutable() : null;
+
+            internal bool SetCurrentCallable(QsQualifiedName name)
+            {
+                this.CurrentCallable = name;
+                return name != null && this.Identifiers.TryAdd(name, ImmutableHashSet.CreateBuilder<string>());
+            }
+
+            internal bool AddIdentifier(string id)
+            {
+                var cannotSet = this.CurrentCallable == null || id == null;
+                var exists = this.Identifiers.TryGetValue(this.CurrentCallable, out var ids);
+                if (cannotSet || !exists) return false;
+                else ids.Add(id);
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Helper class for ListIdentifiers that defines namespace transformations.
```

Comment:
This is another use of 'helper class' that could maybe be rephrased.

Reply:
done